src/solvers/dg_curved/containers_3d.jl

```diff
@@ -68,25 +67,83 @@ function calc_jacobian_matrix!(jacobian_matrix::AbstractArray{<:Any,6}, element,
 end
 
 
-# Calculate contravarant vectors, multiplied by the Jacobian determinant J of the transformation mapping.
-# Those are called Ja^i in Kopriva's blue book.
-function calc_contravariant_vectors!(contravariant_vectors::AbstractArray{<:Any,6}, element, jacobian_matrix)
-  # TODO: This needs to be adapted for actually curved meshes.
-  # For rectangular meshes, the contravariant_vectors are just the scaled unit vectors
-  fill!(view(contravariant_vectors, .., element), 0)
+# Calculate contravariant vectors, multiplied by the Jacobian determinant J of the transformation mapping,
+# using the invariant curl form.
+# These are called Ja^i in Kopriva's blue book.
+function calc_contravariant_vectors!(contravariant_vectors::AbstractArray{<:Any,6}, element, 
+                                     jacobian_matrix, node_coordinates, basis::LobattoLegendreBasis)
+  # The general form is
+  # Jaⁱₙ = 0.5 * ( ∇ × (Xₘ ∇ Xₗ - Xₗ ∇ Xₘ) )ᵢ  where (n, m, l) cyclic and ∇ = (∂/∂ξ, ∂/∂η, ∂/∂ζ)ᵀ
+
+  # Calculate the first summand of the cross product in each dimension
+  for n in 1:3, j in eachnode(basis), i in eachnode(basis)
+    # (n, m, l) cyclic
+    m = (n % 3) + 1
+    l = ((n + 1) % 3) + 1
+
+    # Calc only the first summand 0.5 * (Xₘ Xₗ_ζ - Xₗ Xₘ_ζ)_η of
+    # Ja¹ₙ = 0.5 * [ (Xₘ Xₗ_ζ - Xₗ Xₘ_ζ)_η - (Xₘ Xₗ_η - Xₗ Xₘ_η)_ζ ]
+    @views mul!(contravariant_vectors[1, n, i, :, j, element],
+                basis.derivative_matrix,
+                0.5 * (node_coordinates[m, i, :, j, element] .* jacobian_matrix[l, 3, i, :, j, element] .-
+                       node_coordinates[l, i, :, j, element] .* jacobian_matrix[m, 3, i, :, j, element]))
```

Comment:
You're still allocating for each iteration but that's probably not important here since this function is not performance-critical.

Reply:
I tried to avoid allocations by using `mul!`. I tried to find a non-allocating way to write the multiplication of `node_coordinates` and `jacobian_matrix`, but I didn't find one that doesn't make this less readable than it already is.

In fact, I just did a few benchmarks, and it turns out that the version with `mul!` actually allocates more (and is a little slower) than a simple multiplication.
I decided to use a simple multiplication instead now, since it's more readable and even a bit faster.

Reply:
Sounds ok to me... feel free to mark this as resolved